src/zimscraperlib/zim/dedup.py

```diff
@@ -0,0 +1,73 @@
+import pathlib
+import re
+from typing import Any
+
+import xxhash
+
+from zimscraperlib.zim.creator import Creator
+
+
+class Deduplicator:
+    """Automatically deduplicate potential ZIM items before adding them to the ZIM
+
+    This class automatically computes the digest of every item added to the ZIM, and
+    either add the entry (if item is not yet inside the ZIM) or a redirect (if item with
+    same digest has already been added inside the ZIM).
+
+    This class must be configured with filters to specifiy which items paths to
+    consider. It is of course possible to consider all paths (i.e. all items) with a
+    wide regex or to operate on a subset (e.g. all images) with more precise filters.
+
+    Only the digest/path of items matching the filters are computed and stored.
+
+    The xxh32 algorithm which is known to be good at avoiding collision with minimal
+    memory and CPU footprint is used, so the sheer memory consumption will come from the
+    paths we have to keep. This hashing algorithm is not meant for security purpose
+    since one might infer original content from hashes, but this is not our use case.
+    """
+
+    def __init__(self):
+        self.filters: list[re.Pattern[str]] = []
+        self.added_items: dict[bytes, str] = {}
+
+    def add_item_or_redirect(
+        self,
+        creator: Creator,
+        path: str,
+        *,
+        fpath: pathlib.Path | None = None,
+        content: bytes | str | None = None,
+        **kwargs: Any,
+    ):
+        """Add an item at given path or a redirect"""
+        existing_item = None
+        if any(_filter.match(path) is not None for _filter in self.filters):
+            if content:
+                digest = xxhash.xxh32(
+                    content.encode() if isinstance(content, str) else content
+                ).digest()
+            else:
+                if not fpath:
+                    raise Exception("Either content or fpath are mandatory")
+                xxh32 = xxhash.xxh32()
+                with open(fpath, "rb") as f:
+                    while True:
+                        data = f.read(65536)  # lets read stuff in 64kb chunks
+                        if not data:
+                            break
+                        xxh32.update(data)
+                digest = xxh32.digest()
+
+            if existing_item := self.added_items.get(digest):
+                creator.add_redirect(
```

Comment:
We should use aliases. We have no control over how those items are meant to be used and redirects handling is not transparent. Alias on the other end is.

Reply:
Fine, done